File: todo/usecase/todo.go
package usecase

import (
	"encoding/json"
	"log"
	"todo/models"
	"todo/repository"
)

type todoUsecase struct {
	todoRepository repository.TodoRepository
}

type TodoUsecase interface {
	CreateTodo(todo models.Todo) (bool, error)
	GetTodos(userID int64) (string, error)
	GetTodo(userID, itemID int64) (string, error)
	UpdateTodo(todo models.Todo) (bool, error)
	DeleteTodo(id int64) (bool, error)
}

func InitUserUsecase(todoRepository repository.TodoRepository) TodoUsecase {
	return &todoUsecase{
		todoRepository,
	}
}

func (todoUsecase *todoUsecase) CreateTodo(todo models.Todo) (bool, error) {
	_, err := todoUsecase.todoRepository.CreateTodo(todo)
	if err != nil {
		log.Println("Error create todo", err)
		return false, err
	}

	return true, nil
}

func (todoUsecase *todoUsecase) GetTodos(userID int64) (string, error) {
	todos, err := todoUsecase.todoRepository.GetTodos(userID)
	if err != nil {
		log.Println("Error get todos", err)
		return "", err
	}

	result, err := json.Marshal(todos)

	return string(result), nil
}

func (todoUsecase *todoUsecase) GetTodo(itemID, userID int64) (string, error) {
	todo, err := todoUsecase.todoRepository.GetTodo(itemID, userID)
	if err != nil {
		log.Println("Error get todos", err)
		return "", err
	}

	result, err := json.Marshal(todo)

	return string(result), nil
}

func (todoUsecase *todoUsecase) UpdateTodo(todo models.Todo) (bool, error) {
	_, err := todoUsecase.todoRepository.UpdateTodo(todo)
	if err != nil {
		log.Println("Error update todo", err)
		return false, err
	}

	return true, nil
}

func (todoUsecase *todoUsecase) DeleteTodo(id int64) (bool, error) {
	_, err := todoUsecase.todoRepository.DeleteTodo(id)
	if err != nil {
		log.Println("Error delete todo", err)
		return false, err
	}

	return true, nil
}

File: main/todo/todo.py
import json
from flask import (
    Blueprint,
    g,
    jsonify,
    request,
    render_template,
    session,
    redirect,
    url_for,
)

from todo.client import TodoClient
from middleware.middleware import token_required

todo_blueprint = Blueprint("todo_blueprint", __name__)


@todo_blueprint.route("/create", methods=["GET", "POST"])
@token_required
def create_todo():
    if request.method == "POST":
        try:
            title = request.form["title"]
            description = request.form["description"]
            user_data = session["user_data"]
            if title == "" or description == "":
                raise ValueError("Please fill the title and description")
            client = TodoClient()
            client.create_todo(
                title=request.form["title"],
                description=description,
                user_id=user_data.get("ID"),
            )
            return redirect(url_for("todo_blueprint.get_todos"))
        except Exception as error:
            return render_template("todo/create.html", error=str(error))
    return render_template("todo/create.html")


@todo_blueprint.route("/get", methods=["GET"])
@token_required
def get_todos():
    try:
        user_data = session["user_data"]
        client = TodoClient()
        result = client.get_todos(user_id=user_data.get("ID"))
        data = json.loads(result.data)
        return render_template("todo/list.html", todos=data)
    except Exception as error:
        return render_template("todo/list.html", error=str(error))


@todo_blueprint.route("/get/<int:todo_id>", methods=["GET"])
@token_required
def get_todo(todo_id):
    try:
        user_data = session["user_data"]
        client = TodoClient()
        result = client.get_todo(item_id=todo_id, user_id=user_data.get("ID"))
        data = json.loads(result.data)
        return render_template("todo/item.html", todo=data)
    except Exception as error:
        return render_template("todo/item.html", error=str(error))


@todo_blueprint.route("/update/<int:todo_id>", methods=["POST"])
@token_required
def update_todo(todo_id):
    try:
        user_data = session["user_data"]
        client = TodoClient()
        client.update_todo(todo_id, request.form["title"], request.form["description"])
        result_get = client.get_todo(todo_id, user_data.get("ID"))
        json.loads(result_get.data)
        return redirect(url_for("todo_blueprint.get_todos"))
    except Exception as error:
        return render_template("todo/item.html", error=str(error))


@todo_blueprint.route("/delete/<int:todo_id>", methods=["GET"])
@token_required
def delete_todo(todo_id):
    try:
        client = TodoClient()
        client.delete_todo(todo_id)
        return redirect(url_for("todo_blueprint.get_todos"))
    except Exception as error:
        return render_template("todo/item.html", error=str(error))
